File: java/design_pattern/Strategy.java
import java.text.ParseException;
import java.util.Arrays;
import java.util.List;

public class Strategy {
    public static void main(String[] args) throws ParseException {
        final String d = "a,b,c,d,f,g,e";
        final ExContext ec = new ExContext(d);

        System.out.println(d.getClass());
        System.out.println(d);

        System.out.println(ec.getData().getClass());
        System.out.println(ec.getData());
    }
}

class ExContext {
    private String data;
    private ParserStrategy parserStrategy;

    ExContext(String data) {
        this.data = data;

        if (data.startsWith("{") || data.startsWith("[")) {
            this.parserStrategy = new JsonParser();
        }

        this.parserStrategy = new CsvParser();
    }

    List<String> getData() throws ParseException {
        return this.parserStrategy.parse(this.data);
    }
}

interface ParserStrategy {
    List<String> parse(String data) throws ParseException;
}

class CsvParser implements ParserStrategy {

    @Override
    public List<String> parse(String data) {
        return Arrays.asList(data.split(","));
    }
}

class JsonParser implements ParserStrategy {

    @Override
    public List<String> parse(String data) throws ParseException {
        // Json Parser 구현
        throw new ParseException("JSON Parser가 구현되어 있지 않습니다.", -1);
    }
}

File: js/design_pattern/strategy.js
const ExContext = class {
  constructor(data) {
    let strategy;
    if (data.startsWith('{') || data.startsWith('[')) {
      strategy = new JsonParser();
    } else {
      strategy = new CsvParser();
    }

    Object.assign(this, {
      data, strategy
    })
  }

  getData() {
    return this.strategy.parse(this.data);
  }
};

const ParserStrategy = class {
  parse(data) {
    throw new Error('Must be override');
  }
};

const CsvParser = class extends ParserStrategy {
  parse(data) {
    return data.split(',');
  }
};

const JsonParser = class extends ParserStrategy {
  parse(data) {
    return JSON.parse(data);
  }
};

(_ => {
  const csv = 'a,b,c,d,f,g,e';
  const ec1 = new ExContext(csv);
  console.log(csv);
  console.log(ec1.getData());

  const json = '["a","b","c","d","f","g","e"]';
  const ec2 = new ExContext(json);
  console.log(json);
  console.log(ec2.getData())
})();